### PR TITLE
Remove Dereferences On "this" Program Variables

### PR DESCRIPTION
**Pull request by matjin**
Hotfix to prevent null false positive null dereferences on the "this" program variable.

### CODE REVIEW

**Review by xi-liu-ds — COMMENTED** [review]
Could you add an unit test for this case? Thanks

**Comment by matjin** [comment]
> Could you add an unit test for this case? Thanks

Validating this isn't supported in the current test framework, as we so far only validate insofar as a null dereference warning can be produced -- it would require us to validate actual CFGs and the instructions which comprise them.

**Comment by xi-liu-ds** [comment]
> > Could you add an unit test for this case? Thanks
> 
> Validating this isn't supported in the current test framework, as we so far only validate insofar as a null dereference warning can be produced -- it would require us to validate actual CFGs and the instructions which comprise them.

Can we construct a dereference of `this` in `TestClass.cs` and check if null dereference will be reported?

**Comment by matjin** [comment]
> > > Could you add an unit test for this case? Thanks
> > 
> > 
> > Validating this isn't supported in the current test framework, as we so far only validate insofar as a null dereference warning can be produced -- it would require us to validate actual CFGs and the instructions which comprise them.
> 
> Can we construct a dereference of `this` in `TestClass.cs` and check if null dereference will be reported?

It's very rare for infer to report a null dereference on "this" (Infer actually had to have thought "this" was a null reference) and I'm actually not sure how to reproduce it -- it seems to be an odd/rare edge case that appeared in some of the benchmarking, where there was null k__BackingField and infer associated "this" with that for some reason.

File: Cilsil/Cil/Parsers/LeaveParser.cs [hunk 1]
```diff
@@ -20,9 +20,20 @@ internal class LeaveParser : InstructionParser
                 // Leave for catch block: target instruction immediately after catch block.
                 case Code.Leave:
                 case Code.Leave_S:
+                case Code.Throw:
```
Comment (xi-liu-ds) [inline Cilsil/Cil/Parsers/LeaveParser.cs:23]: Why `Throw` is handled in `LeaveParser`? Didn't we have a `ThrowParser` already?

**Review by xi-liu-ds — COMMENTED** [review]
I think this PR is generally for exception handling, not really specifically for race condition detection. Do you want to redirect this PR to your exception handling branch?

**Comment by xi-liu-ds** [comment]
> > > > Could you add an unit test for this case? Thanks
> > > 
> > > 
> > > Validating this isn't supported in the current test framework, as we so far only validate insofar as a null dereference warning can be produced -- it would require us to validate actual CFGs and the instructions which comprise them.
> > 
> > 
> > Can we construct a dereference of `this` in `TestClass.cs` and check if null dereference will be reported?
> 
> It's very rare for infer to report a null dereference on "this" (Infer actually had to have thought "this" was a null reference) and I'm actually not sure how to reproduce it -- it seems to be an odd/rare edge case that appeared in some of the benchmarking, where there was null k__BackingField and infer associated "this" with that for some reason.

If then, I think I need to benchmark this PR to see if there is any side effect.

**Comment by matjin** [comment]
> I think this PR is generally for exception handling, not really specifically for race condition detection. Do you want to redirect this PR to your exception handling branch?

True -- I probably should have based it off of that. I think I did this in case it created silly merge conflicts with the other downstream branches (since there are a few now). If it's OK, I'd like to go ahead and just merge this for simplicity's sake since it's already set up.

File: Cilsil/Cil/Parsers/RetParser.cs [hunk 2]
```diff
@@ -22,13 +22,18 @@ internal class RetParser : InstructionParser
                     var retType = state.Method.ReturnType.GetElementType();
                     if (retType == state.Method.Module.TypeSystem.Void)
                     {
-                        state.PreviousNode.Successors.Add(state.ProcDesc.ExitNode);
+                        var retNode = new StatementNode(state.CurrentLocation,
```
Comment (xi-liu-ds) [inline Cilsil/Cil/Parsers/RetParser.cs]: Is this fixing anything? Otherwise please remove it.

**Review by xi-liu-ds — APPROVED** [review]

**Pull request merged**